### ISSUE TITLE
Report the case where overloads use `@staticmethod` or `@classmethod` inconsistently

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#6592]
Mypy reports an error when `@staticmethod` or `@classmethod` are used inconsistently in an overloaded function or method. Pyright should do the same.

```python
from typing import overload, Callable, Any

class MyClass:
    @overload
    def connect(
        self, arg__1: bytes, arg__2: Callable[..., Any], /, type: str = ...
    ) -> str: ...
    @overload
    @staticmethod
    def connect(
        arg__1: MyClass, arg__2: bytes, arg__3: Callable[..., Any], /, type: str = ...
    ) -> str: ...
```

This is related to [this issue](https://github.com/microsoft/pyright/issues/6585).

**Comment by pakrym-stripe** [comment]
There are cases where inconsistent overloads describe the API surface correctly. stripe-python allows calling the same method in both a static and instance context:

```python
Account.delete("acc_123")
```

```python
acc = Account.retrieve("acc_123")
acc.delete()
```

A mix of `@staticmethod` and normal overloads is the only way we found to type this API.
 
https://github.com/stripe/stripe-python/blob/master/stripe/api_resources/account.py#L3523-L3547

**Comment by erictraut** [comment]
Other Python type checkers don't support the approach you're using, and I don't think it is supported by the type system. 

I recommend that you post a question in the [Python Typing forum](https://discuss.python.org/c/typing/32). This forum sees participation from the maintainers of all the major Python type checkers, typeshed stubs, the newly-formed [Typing Council](https://peps.python.org/pep-0729/), and core Python developers. Perhaps someone there has a suggestion for how to model this in a way that works within the current type system. If not, this is also the place to discuss extending the type system to accommodate this use case.

**Comment by pakrym-stripe** [comment]
Fair enough, thank you @erictraut!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.339, which I just published. It will also be included in a future release of pylance.
